tracing-core/src/dispatcher.rs

```diff
@@ -669,8 +669,13 @@ impl Drop for DefaultGuard {
     #[inline]
     fn drop(&mut self) {
         if let Some(dispatch) = self.0.take() {
-            let _ = CURRENT_STATE.try_with(|state| {
-                *state.default.borrow_mut() = dispatch;
+            // Replace the dispatcher and then drop the old one outside
+            // of the thread-local context. Dropping the dispatch may
+            // lead to the drop of a subscriber which, in the process,
+            // could then also attempt to access the same thread local
+            // state -- causing a clash.
+            let _prev = CURRENT_STATE.try_with(|state| {
+                state.default.replace(dispatch);
```

Comment:
@d-e-s-o the clippy lint you mentioned is actually showing us a bug here! This doesn't actually make the expected change. Because of the `;` at the end of this line, we are still immediately dropping the previous value, rather than returning it from the closure. As clippy pointed out, the value that was being dropped was a `Result<(), AccessError>`, not a `Result<Dispatch, ...>`. 

So, I think this needs to be
```suggestion
                state.default.replace(dispatch)
```

Reply:
(and then, we can probably put the explicit drop back in if you think it's clearer...)

Reply:
Good that we talked about it then ;) Thanks.

Reply:
Done.